apache/nuttx
# riscv/common/espressif: Fix spi slave driver

### Summary

In the current version, the ESP32-C6 does not behave correctly when sending data while operating in SPI Slave mode. This commit addresses the issue by ensuring proper data transmission. Additionally, Espressif's SPI controller uses a single buffer functioning as both TX and RX, which differs from other MCU manufacturers such as ST and Silicon Labs, as discussed and illustrated in the images below. With this commit, the SPI Slave behavior is expected to be appropriate on Espressif's RISC-V devices.

### Impact

This change affects the operation of SPI Slave mode on Espressif's ESP32-C6 devices. Users utilizing SPI Slave configurations will be able to send data consistently and as expected, aligning with the behavior of SPI controllers from other MCU manufacturers. This improvement enhances interoperability and reliability of SPI communications in projects relying on these devices. Furthermore, the unified TX/RX buffer simplifies maintenance and ensures compatibility with future firmware and hardware updates.

For more details about compatibility, please take a look on https://github.com/apache/nuttx/issues/13855
You will see that the Espressif SPI controller has only one buffer to TX and RX data, so is necessary driver clear buffer if it have no data to be sent.
Below you can see behavior from stm32f4 mcu: 
![image](https://github.com/user-attachments/assets/d6132dc0-34b3-49f4-859a-678dcb20cc42)

And below you can see behavior from efm32gg12 mcu:
![image](https://github.com/user-attachments/assets/5c1da73c-6de3-4f47-a959-28e82844d1c1)
![image](https://github.com/user-attachments/assets/eda27d09-f6af-460c-864e-a8b1438fb9fa)

### Testing

To verify the effectiveness of this change, the spislv configuration available for the esp32c6-devkitm and esp32c6-devkitc kits was used. The tests involved sending data via SPI and observing that the master only receives the data being sent by the slave. The following environments and tools were utilized during testing:

    Build Host: [Specify the operating system, CPU, compiler, etc.]
    Verification Target: Espressif's RISC-V devices, specifically the esp32c6-devkitm and esp32c6-devkitc kits.
    Testing Procedure:
        Configure the mentioned kits in SPI Slave mode using the spislv configuration.
        Send a sequence of data via SPI from the master.
        Verify on the master that only the data sent by the slave is received correctly.
    Results: After the commit, the data sent by the slave was correctly received by the master as expected, without inconsistencies or data loss.

Build and runtime logs before and after the change were analyzed to ensure that the SPI Slave behavior meets the expected outcomes.

Se results below:
![image](https://github.com/user-attachments/assets/f89b8f9b-e4dc-4506-82a1-d150992a7085)

Below I fixed only 2 bytes as answer:
![image](https://github.com/user-attachments/assets/3eacc97d-ea13-4066-b48c-7e233db5872f)
Now only two bytes are send.
